pandas/core/internals/blocks.py

```diff
@@ -2978,6 +2978,29 @@ def where(self, other, cond, align=True, errors='raise',
                                                axis=axis, transpose=transpose)
         return result
 
+    def replace(self, to_replace, value, inplace=False, filter=None,
+                regex=False, convert=True):
+        if filter is None or not self.mgr_locs.isin(filter).any():
```

Comment:
this block looks unrelated to the linked issue.  what case is it solving?  needs a separate test?

Reply:
The first case is used when the replace function applies to the entire block, so the replace can be carried out just by manipulating categories. If there's a filter, this might not be possible so we fix the categories then use the original implementation. Most calls to `Series.replace` should be covered with this case (including the new test in `series/test_replace.py`)

I added a test in `frame/test_replace.py` to cover the second case, where a filter is used. It improves upon the previous behaviour, which would have casted the frame to `object`